src/julia.h

```diff
@@ -2034,7 +2034,10 @@ typedef struct {
 #define jl_root_task (jl_get_ptls_states()->root_task)
 
 // codegen interface ----------------------------------------------------------
-
+// The root propagation here doesn't have to be literal, but callers should
+// ensure that the return value outlives the MethodInstance
+typedef jl_value_t *(*jl_codeinstance_lookup_t)(jl_method_instance_t *mi JL_PROPAGATES_ROOT,
+    size_t min_world, size_t max_world);
```

Comment:
I'd be interested in knowing if we can break this assumption (remove the `JL_PROPAGATES_ROOT` attribute). It is required that the result be rooted (because it could hold a couple of the gc roots for the generated code result)

> outlives the MethodInstance

These are typically required to be permanently allocated (so that we can track their edges), so this would imply they cannot be freed?

Reply:
> These are typically required to be permanently allocated

Yes, I just wanted to annotate this from a lifetime perspective here, since we don't want to require that the `mi` literally roots the result value here.